gax/src/main/java/com/google/api/gax/rpc/Watchdog.java

```diff
@@ -0,0 +1,314 @@
+
+      // Only reset the request water mark if there are no outstanding requests.
+      synchronized (lock) {
+        if (state == State.IDLE) {
+          state = State.WAITING;
+          lastActivityAt = clock.millisTime();
+        }
+
+        // Increment the request count without overflow
+        int maxIncrement = Integer.MAX_VALUE - pendingCount;
+        count = Math.min(maxIncrement, count);
+        pendingCount += count;
+      }
+      innerController.request(count);
+    }
+
+    private void onCancel() {
+      error = new CancellationException("User cancelled stream");
+      innerController.cancel();
+    }
+
+    @Override
+    public void onResponseImpl(ResponseT response) {
+      synchronized (lock) {
+        state = State.DELIVERING;
+      }
+
+      outerResponseObserver.onResponse(response);
+
+      synchronized (lock) {
+        pendingCount--;
+        lastActivityAt = clock.millisTime();
+
+        if (autoAutoFlowControl || pendingCount > 0) {
+          state = State.WAITING;
+        } else {
+          state = State.IDLE;
+        }
+      }
+    }
+
+    @Override
+    public void onErrorImpl(Throwable t) {
+      // Overlay the cancellation errors (either user or idle)
+      if (this.error != null) {
+        t = this.error;
+      }
+      openStreams.remove(this);
+      outerResponseObserver.onError(t);
+    }
+
+    @Override
+    public void onCompleteImpl() {
+      openStreams.remove(this);
+      outerResponseObserver.onComplete();
+    }
+
+    /**
+     * Checks if this stream has overrun any of its timeouts and cancels it if it does.
+     *
+     * @return True if the stream was canceled.
+     */
+    boolean cancelIfStale() {
+      Throwable myError = null;
+
+      synchronized (lock) {
+        long waitTime = clock.millisTime() - lastActivityAt;
+
+        switch (this.state) {
+          case IDLE:
+            if (!idleTimeout.isZero() && waitTime >= idleTimeout.toMillis()) {
+              myError = new IdleConnectionException("Canceled due to idle connection", false);
+            }
+            break;
+          case WAITING:
+            if (!waitTimeout.isZero() && waitTime >= waitTimeout.toMillis()) {
+              myError =
+                  new IdleConnectionException(
+                      "Canceled due to timeout waiting for next response", true);
+            }
+            break;
+        }
+      }
+
+      if (myError != null) {
+        this.error = myError;
+        innerController.cancel();
+        return true;
+      }
+      return false;
+    }
+  }
+
+  /** The marker exception thrown when a timeout is exceeded. */
+  public static class IdleConnectionException extends ApiException {
+
+    IdleConnectionException(String message, boolean retry) {
+      super(message, null, LOCAL_ABORTED_STATUS_CODE, retry);
+    }
+  }
+
+  public static final StatusCode LOCAL_ABORTED_STATUS_CODE =
```

Comment:
Minor: can this be package-private?

Reply:
I would like to leave this public to allow users to catch this